src/targets/gpu/write_literals.cpp

```diff
@@ -27,36 +27,155 @@
 #include <migraphx/instruction.hpp>
 #include <migraphx/program.hpp>
 #include <migraphx/env.hpp>
+#include <migraphx/liveness.hpp>
+#include <migraphx/register_op.hpp>
 
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 namespace gpu {
 
 MIGRAPHX_DECLARE_ENV_VAR(MIGRAPHX_COPY_LITERALS)
 
+struct literal_as_argument
+{
+    std::string name() const { return "gpu::literal_as_argument"; }
+    argument data;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.data, "data"));
+    }
+
+    shape compute_shape(const std::vector<shape>&) const { return data.get_shape(); }
+    argument compute(context&, const shape&, const std::vector<argument>&) const { return data; }
+    friend std::ostream& operator<<(std::ostream& os, const literal_as_argument& x)
+    {
+        os << x.name() << "[data=";
+        if(x.compute_shape(std::vector<shape>{}).elements() < 10)
+        {
+            os << x.data;
+        }
+        else
+        {
+            os << "{ ... }";
+        }
+        os << "]";
+        return os;
+    }
+};
+MIGRAPHX_REGISTER_OP(literal_as_argument);
+
 void write_literals::apply(module& m) const
 {
     assert(ctx != nullptr);
     std::size_t n = 0;
-    for(auto ins : iterator_for(m))
+
+    if(weight_streaming)
     {
-        if(ins->name() == "@literal")
+        std::size_t bytes_on_gpu = 0;
+        size_t scratch_size      = 0;
+        liveness(m, [&](auto ins, auto live_set) {
+            if(ins->name() != "hip::allocate" or ins->get_shape().bytes() == 0)
+            {
+                return;
+            }
+            size_t temp_size = 0;
+            for(auto i : live_set)
+            {
+                if(i->name() != "hip::allocate" or i->get_shape().bytes() == 0)
+                {
+                    continue;
+                }
+                temp_size += i->get_shape().bytes();
+            }
+
+            if(temp_size > scratch_size)
+            {
+                scratch_size = temp_size;
+            }
+        });
+
+        std::vector<instruction_ref> ins_list;
+        size_t size_of_literals = 0;
+        for(auto ins : iterator_for(m))
         {
-            if(enabled(MIGRAPHX_COPY_LITERALS{}))
+            if(ins->name() == "@literal")
+            {
+                ins_list.push_back(ins);
+                size_of_literals += ins->get_shape().bytes();
+            }
+        }
+
+        long budget = max_memory;
+        if(budget == LONG_MAX)
+        {
+            budget = static_cast<long>(size_of_literals / 4);
+        }
+
+        size_t free_memory = 0;
```

Comment:
This value is never used. It should be used to set the `budget` when its not been set.

Reply:
Changed, thanks